### PR TITLE
New Keybindings Refactor

### PR DESCRIPTION
**Pull request by blubaugha**
# Attention!
I left the current **keybinding** code in place (but disabled it), and only added and modified code, to make it easier to see what's changed. The next PR will be to remove the current **keybinding** code.

### Summary
This is a rewrite of the way client-side **keybindings** work in **JumpMind Commerce**. Things are more automatic and "just work" without the need to manually write code for components/directives and various problems/hacks are no more.

### What's Different

* There are now "**Keybinding Zones**" and only one can be active at a time
  * Examples of **Keybinding Zones**: The active screen or dialog, a kebab menu, bacon strip etc.
* Keybinding are automatically registered and there's no need to manually write code to wire them up
  * Just add actions to your `Java` objects and they'll work on the client side
  * Keybindings will automatically work for new screens and dialogs (but you can write custom code if you need to)
* Actions are automatically executed when a keybinding is pressed 
* The main keybinding service is now a singleton (makes it easier to enforce only one active zone)
* All keyboard events are allowed to bubble
  * This allows the browser's default behavior to work for things like pressing enter in a dialog always submits without any hacks (currently not working correctly)
* Keypress events are "captured" on the way down the DOM (instead of when they bubble up) so that nobody can stop propagation and prevent a key press from being seen by new service
* No more random problems where keyboard keys randomly stop working
  * Key events are monitored at the window, instead of in different sections in the page, to ensure no events are missed.
* The keybindings are disabled when the screen is locked, and the default browser behavior works with the `Enter` key to submit without the need for any hacks (because events are allowed to bubble)
* A _mOrE_ better separation of concerns

### Scenarios I Tested
There's probably some spots I missed, but the core things are working with keybindings:

* Home screen keybindings
* Toggling the bacon strip (`Escape` key)
* Executing bacon strip actions by directly pressing keybinding key and navigating with arrow keys and pressing `Enter`
* Executing kebab actions
* Forms/dialogs - `Enter` key submits forms
* Toggling the Loyalty Lookup (`F4` to open,`Escape` to close and `Enter` to submit search)
* Arrow navigation in bacon strip, sale items and kebab menu, tendering dialogs
* Navigating between pages, dialogs, kebab menus

### A Couple of Notes
* It's easier to see how some of the usage has changed by ignoring all of the added files and only looking at the modified files.
* The `KeybindingParserService` is a port of the relevant keybinding logic in the existing `KeyPressProvider` service.

### Code Responsibilities

`KeybindingService`
The core (singleton) service responsible for the keybinding functionality. This ensures only on one keybinding zone is active at a time and automatically executes actions for their corresponding keybindings.

`KeybindingLockScreenService`
Manages enabling/disabling keybindings when the screen is locked/unlocked

`KeybindingParserService`
Parses keybinding strings, like `Enter`, `Ctrl+Shift+F10` and `ArrowUp,ArrowDown` etc.

`KeybindingPropertyCrawlerService`
Crawls arbitrary object hierarchies and finds all actions with keybindings

`KeybindingZoneService`
Used by UI components/directives to interact with the current **KeybindingZone**. Each instance is "locked" to a specific zone id (usually the screen or dialog message id).

`KeybindingZoneScreenService`
Manages the automatic activating/deactivating **KeybindingZone** for screens and dialogs

### CODE REVIEW

File: openpos-client-libs/projects/openpos-client-core-lib/src/lib/core/keybindings/keybinding-lock-screen.service.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            MessageTypes.UNLOCK_SCREEN
+        ).pipe(
+            takeUntil(this.destroyed$)
+        ).subscribe(message => this.handleLockScreenMessage(message));
+
+        console.log('[KeybindingLockScreenService]: Listening for message types "LockScreen" and "UnlockScreen"');
+    }
+
+    handleLockScreenMessage(message: any): void {
+        if (message.type === MessageTypes.LOCK_SCREEN) {
```
[the remark this answers is not in the document]
Comment (blubaugha) [reply openpos-client-libs/projects/openpos-client-core-lib/src/lib/core/keybindings/keybinding-lock-screen.service.ts:30]: Ohhhhh, good test - I didn't think about that. 

Things did work correctly though. 

It looks like the server correctly sends down the lock/unlock message.

**Review by kdelorey — COMMENTED** [review]
Out of curiosity, have you looked into unit testing the zone stuff? I'm not sure what it looks like, especially with the keyboard in play, but I feel like it would give me warm fuzzies about its future.

**Comment by blubaugha** [comment]
> Out of curiosity, have you looked into unit testing the zone stuff? I'm not sure what it looks like, especially with the keyboard in play, but I feel like it would give me warm fuzzies about its future.

Yes, tests are on the way!

But I'm currently stuck trying to figure out why I can't get the Karma tests to run in Intellij...

**Comment by blubaugha** [comment]
@kdelorey 
I added unit tests and have the most important parts covered
Screen Shot 2022-01-06 at 5 47 09 PM

**Comment by kdelorey** [comment]
Your tests and coverage gives me warm fuzzies now. You're the man!

As an aside, there's a new process that we're going to start rolling out I think and I think this issue is a good candidate to start transitioning to the process. The idea is to support a better form of release notes by creating a paired issue for the PR.

I know this process was started after this PR was created, but would you mind creating a paired issue (maybe bug? that's what initiated this work right?) that describes the reason for this PR?

**Comment by blubaugha** [comment]
> Your tests and coverage gives me warm fuzzies now. You're the man!
> 
> As an aside, there's a new process that we're going to start rolling out I think and I think this issue is a good candidate to start transitioning to the process. The idea is to support a better form of release notes by creating a paired issue for the PR.
> 
> I know this process was started after this PR was created, but would you mind creating a paired issue (maybe bug? that's what initiated this work right?) that describes the reason for this PR?

An issue has been linked captain!
https://github.com/JumpMind/openpos-framework/issues/1874

**Review by kdelorey — DISMISSED** [review]

**Review by kdelorey — DISMISSED** [review]

**Review by kdelorey — DISMISSED** [review]

**Review by kdelorey — DISMISSED** [review]

**Review by kdelorey — DISMISSED** [review]

**Comment by blubaugha** [comment]
@kdelorey 
I finally have really good coverage here!

**Review by eddielegg-le — APPROVED** [review]

**Pull request merged**